bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -2335,36 +2335,30 @@ private static boolean checkIsLikeArrayType(Object sourceValue, BArrayType targe
 
         ArrayValue source = (ArrayValue) sourceValue;
         Type targetTypeElementType = targetType.getElementType();
+
         if (source.getType().getTag() == TypeTags.ARRAY_TAG) {
             Type sourceElementType = ((BArrayType) source.getType()).getElementType();
             if (isValueType(sourceElementType)) {
                 boolean isType = checkIsType(sourceElementType, targetTypeElementType, new ArrayList<>());
 
-                if (isType || !allowNumericConversion || !isNumericType(sourceElementType)) {
+                if (isType || !isNumericType(sourceElementType)) {
                     return isType;
                 }
 
-                if (isNumericType(targetTypeElementType)) {
-                    return true;
-                }
-
-                if (targetTypeElementType.getTag() != TypeTags.UNION_TAG) {
-                    return false;
-                }
-
-                List<Type> targetNumericTypes = new ArrayList<>();
-                for (Type memType : ((BUnionType) targetTypeElementType).getMemberTypes()) {
-                    if (isNumericType(memType) && !targetNumericTypes.contains(memType)) {
-                        targetNumericTypes.add(memType);
+                if (targetTypeElementType.getTag() == TypeTags.UNION_TAG) {
+                    List<Type> targetNumericTypes = new ArrayList<>();
+                    for (Type memType : ((BUnionType) targetTypeElementType).getMemberTypes()) {
+                        if (isNumericType(memType) && !targetNumericTypes.contains(memType)) {
+                            targetNumericTypes.add(memType);
+                        }
                     }
+                    return targetNumericTypes.size() == 1;
```

Comment:
Shouldn't we return true here only if allowNumericConversion is also true?

Reply:
Please consider the following example.

```ballerina
import ballerina/io;

public function main() {
    int[] & readonly x = [1, 2];
    any a = x;

    io:println(a is (boolean|float)[]); // Should print `false`, prints `true` with this PR.
}
```

Reply:
Yes, it should happen only if the `allowNumericConversion` is true and the source element type is a numeric type. Misunderstood the logic before and corrected with the latest commits.